mimic/apache.go

```diff
@@ -0,0 +1,321 @@
+	}
+	path := req.URL.Path
+	// remove extra leading slash
+	if len(path) > 0 && path[0] == '/' {
+		i := 1
+		for ; i < len(path) && path[i] == '/'; i++ {
+		}
+		if i > 0 {
+			i--
+		}
+		path = path[i:]
+	}
+	m := apacheMimic{conn, req, path}
+	if req.Host == "" {
+		m.writeError(badRequestHeader, badRequestBody)
+		return
+	}
+	switch req.Method {
+	case "CONNECT":
+		m.writeError(badRequestHeader, badRequestBody)
+	case "GET", "POST":
+		switch path {
+		case "/", "/index.html":
+			m.ok(indexHeader, indexDotHTML)
+		case "/icons/ubuntu-logo.png":
+			m.ok(logoHeader, ubuntuLogo)
+		default:
+			m.writeError(notFoundHeader, notFoundBody)
+		}
+	case "HEAD":
+		switch path {
+		case "/", "/index.html":
+			m.ok(indexHeader, nil)
+		case "/icons/ubuntu-logo.png":
+			m.ok(logoHeader, nil)
+		default:
+			m.writeError(notFoundHeaderWhenHead, nil)
+		}
+	case "OPTIONS":
+		switch path {
+		case "/", "/index.html":
+			m.ok(optionsHeader, nil)
+		case "/icons/ubuntu-logo.png":
+			m.ok(optionsHeader, nil)
+		default:
+			m.writeError(optionsHeaderWhenNotFound, nil)
+		}
+	default:
+		if m.path == "/" {
+			m.path = "/index.html"
+		}
+		switch {
+		case !KNOWN_METHODS[req.Method]:
+			m.writeError(notImplementedHeader, notImplementedBody)
+		case !ALLOWED_METHODS[req.Method]:
+			m.writeError(methodNotAllowedHeader, methodNotAllowedBody)
+		case KNOWN_URIS[path]:
+			//m.ok()
+		default:
+			m.writeError(notFoundHeader, notFoundBody)
+		}
+	}
+}
+
+func (f *apacheMimic) ok(header *template.Template, body []byte) {
+	var buf bytes.Buffer
+	err := header.Execute(&buf, f.collectVars())
+	if err != nil {
+		panic(fmt.Sprintf("execute template err: %s", err))
+	}
+	// ignore any errors writing back to connection
+	_, _ = buf.WriteTo(f.conn)
+	_, _ = f.conn.Write(body)
+	_ = f.conn.Close()
+}
+
+func (f *apacheMimic) writeError(header, body *template.Template) {
+	var headBuf bytes.Buffer
+	var bodyBuf bytes.Buffer
+	vars := f.collectVars()
+	if body != nil {
+		err := body.Execute(&bodyBuf, vars)
+		if err != nil {
+			panic(fmt.Sprintf("execute template err: %s", err))
+		}
+		vars.ContentLength = bodyBuf.Len()
+	}
+	err := header.Execute(&headBuf, vars)
+	if err != nil {
+		panic(fmt.Sprintf("execute template err: %s", err))
+	}
+	// ignore any errors writing back to connection
+	_, _ = headBuf.WriteTo(f.conn)
+	_, _ = bodyBuf.WriteTo(f.conn)
+	_ = f.conn.Close()
+}
+
+type vars struct {
+	Date, LastModified, ETag, Path, Host, Port string
+	ContentType                                string
+	ContentLength                              int
+}
+
+func (f *apacheMimic) collectVars() *vars {
+	return &vars{
+		Date:         time.Now().Format(timeFormat),
+		LastModified: lastModified,
+		ETag:         etag,
+		Path:         f.path,
+		Host:         Host,
+		Port:         Port,
+	}
+}
+
+func makeETag() string {
+	const alphanum = "0123456789abcdefghijklmnopqrstuvwxyz"
+	bytes := [17]byte{}
+	rand.Read(bytes[:])
+	for i, b := range bytes {
```

Comment:
Sorry for the nitpicking... why don't you run the second case and outside the range looping just set:
```
bytes[4] = '-'
```